src/pm/xbps.rs

```diff
@@ -0,0 +1,241 @@
+#![doc = docs_self!()]
+
+use async_trait::async_trait;
+use futures::prelude::*;
+use indoc::indoc;
+use once_cell::sync::Lazy;
+use tap::Pipe;
+use tokio::io::AsyncWriteExt;
+
+use super::{Pm, PmHelper, PmMode, PromptStrategy, Strategy};
+use crate::{
+    dispatch::Config,
+    error::{Error, Result},
+    exec::Cmd,
+};
+
+macro_rules! docs_self {
+    () => {
+        indoc! {"
+            The [X Binary Package System](https://github.com/void-linux/xbps)
+        "}
+    };
+}
+
+#[doc = docs_self!()]
+#[derive(Debug)]
+pub(crate) struct Xbps {
+    cfg: Config,
+}
+
+static STRAT_PROMPT: Lazy<Strategy> = Lazy::new(|| Strategy {
+    prompt: PromptStrategy::native_no_confirm(&["--yes"]),
+    ..Strategy::default()
+});
+
+impl Xbps {
+    #[must_use]
+    #[allow(missing_docs)]
+    pub(crate) fn new(cfg: Config) -> Self {
+        Xbps { cfg }
+    }
+}
+
+#[async_trait]
+impl Pm for Xbps {
+    /// Gets the name of the package manager.
+    fn name(&self) -> &str {
+        "xbps"
+    }
+
+    fn cfg(&self) -> &crate::dispatch::Config {
+        &self.cfg
+    }
+
+    /// Q generates a list of installed packages.
+    async fn q(&self, kws: &[&str], flags: &[&str]) -> Result<()> {
+        if kws.is_empty() {
+            return self
+                .run(Cmd::new(&["xbps-query", "-l"]).kws(kws).flags(flags))
+                .await;
+        }
+
+        let lines: Vec<_> = stream::iter(kws)
+            .map(Ok)
+            .and_then(|&pkg| async {
+                let cmd = Cmd::new(&["xbps-query", "--property", "pkgver", pkg]).flags(flags);
+                self.check_output(cmd, PmMode::Mute, &Strategy::default())
+                    .await
+                    .map_err(|err| (pkg.to_string(), err))
+            })
+            .collect()
+            .await;
+
+        let mut stdout = tokio::io::stdout();
+        for line in lines {
+            match line {
+                Ok(line) => stdout.write_all(&line).await?,
+                Err((pkg, Error::CmdStatusCodeError { .. })) => {
+                    let msg = format!("error: package '{}' was not found\n", pkg);
```

Comment:
In terms of error recovery on failure, I think you could be even more precise. I guess `xbps` will give a specific error code when a package is not found, so that your error message will be guaranteed to pop up at the right time (I'm really afraid of giving inappropriate instructions to users @_@). And if not, you can still match against the output :)

Reply:
From my manual tests it seems that xbps gives code 2 when it cannot find the queried, I'll match against that

Reply:
Done!